### PR TITLE
Token based authentication extension libraries

### PR DESCRIPTION
**Pull request by atakavci**
This PR contains ;
- the Core(`redis-authx-core`) components for token based authentication as a module to be shared among [jedis](https://github.com/redis/jedis) and [lettuce](https://github.com/redis/lettuce)
- A common implementation of authentication(`redis-authx-entraid`) library to support **MS EntraID** to help support integration of jedis and lettuce with **MS EntraID** authentication 
- Reqired workflows and configuration regarding to maintain and test both  core and entraid modules

Focused aspects are;
- providing possible future extensibility with the core library
- providing generic interfaces/components that allow other identity providers to be implemented.
- keep allowing configurations that MS EntraID support 
- providing easy to use and stable interfaces for particular requirements of individual client library(jedis and lettuce) .

### CODE REVIEW

**Review by ggivo — APPROVED** [review]
Looks good to me.
I think it is good first version

File: core/src/main/java/redis/clients/authentication/core/TokenManagerConfig.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
+package redis.clients.authentication.core;
```
Comment (tishun) [inline core/src/main/java/redis/clients/authentication/core/TokenManagerConfig.java:1]:
```suggestion
/*
 * Copyright 2024, Redis Ltd. and Contributors
 * All rights reserved.
 *
 * Licensed under the MIT License.
 */
package redis.clients.authentication.core;
```

File: core/src/main/java/redis/clients/authentication/core/TokenRequestException.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package redis.clients.authentication.core;
```
Comment (tishun) [inline core/src/main/java/redis/clients/authentication/core/TokenRequestException.java:1]:
```suggestion
/*
 * Copyright 2024, Redis Ltd. and Contributors
 * All rights reserved.
 *
 * Licensed under the MIT License.
 */
package redis.clients.authentication.core;
```

File: core/src/test/java/redis/clients/authentication/CoreAuthenticationUnitTests.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
+package redis.clients.authentication;
```
Comment (tishun) [inline core/src/test/java/redis/clients/authentication/CoreAuthenticationUnitTests.java:1]:
```suggestion
/*
 * Copyright 2024, Redis Ltd. and Contributors
 * All rights reserved.
 *
 * Licensed under the MIT License.
 */
package redis.clients.authentication;
```

File: entraid/src/main/java/redis/clients/authentication/entraid/EntraIDIdentityProvider.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
+package redis.clients.authentication.entraid;
```
Comment (tishun) [inline entraid/src/main/java/redis/clients/authentication/entraid/EntraIDIdentityProvider.java:1]:
```suggestion
/*
 * Copyright 2024, Redis Ltd. and Contributors
 * All rights reserved.
 *
 * Licensed under the MIT License.
 */
package redis.clients.authentication.entraid;
```

File: entraid/src/main/java/redis/clients/authentication/entraid/EntraIDIdentityProviderConfig.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
+package redis.clients.authentication.entraid;
```
Comment (tishun) [inline entraid/src/main/java/redis/clients/authentication/entraid/EntraIDIdentityProviderConfig.java:1]:
```suggestion
/*
 * Copyright 2024, Redis Ltd. and Contributors
 * All rights reserved.
 *
 * Licensed under the MIT License.
 */
package redis.clients.authentication.entraid;
```

File: entraid/src/main/java/redis/clients/authentication/entraid/EntraIDTokenAuthConfigBuilder.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
+package redis.clients.authentication.entraid;
```
Comment (tishun) [inline entraid/src/main/java/redis/clients/authentication/entraid/EntraIDTokenAuthConfigBuilder.java:1]:
```suggestion
/*
 * Copyright 2024, Redis Ltd. and Contributors
 * All rights reserved.
 *
 * Licensed under the MIT License.
 */
package redis.clients.authentication.entraid;
```

File: entraid/src/main/java/redis/clients/authentication/entraid/JWToken.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
+package redis.clients.authentication.entraid;
```
Comment (tishun) [inline entraid/src/main/java/redis/clients/authentication/entraid/JWToken.java:1]:
```suggestion
/*
 * Copyright 2024, Redis Ltd. and Contributors
 * All rights reserved.
 *
 * Licensed under the MIT License.
 */
package redis.clients.authentication.entraid;
```

File: entraid/src/main/java/redis/clients/authentication/entraid/ManagedIdentityInfo.java [hunk 8]
```diff
@@ -0,0 +1,50 @@
+package redis.clients.authentication.entraid;
```
Comment (tishun) [inline entraid/src/main/java/redis/clients/authentication/entraid/ManagedIdentityInfo.java:1]:
```suggestion
/*
 * Copyright 2024, Redis Ltd. and Contributors
 * All rights reserved.
 *
 * Licensed under the MIT License.
 */
package redis.clients.authentication.entraid;
```

File: entraid/src/main/java/redis/clients/authentication/entraid/RedisEntraIDException.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
+package redis.clients.authentication.entraid;
```
Comment (tishun) [inline entraid/src/main/java/redis/clients/authentication/entraid/RedisEntraIDException.java:1]:
```suggestion
/*
 * Copyright 2024, Redis Ltd. and Contributors
 * All rights reserved.
 *
 * Licensed under the MIT License.
 */
package redis.clients.authentication.entraid;
```

File: entraid/src/main/java/redis/clients/authentication/entraid/ServicePrincipalInfo.java [hunk 10]
```diff
@@ -0,0 +1,58 @@
+package redis.clients.authentication.entraid;
```
Comment (tishun) [inline entraid/src/main/java/redis/clients/authentication/entraid/ServicePrincipalInfo.java:1]:
```suggestion
/*
 * Copyright 2024, Redis Ltd. and Contributors
 * All rights reserved.
 *
 * Licensed under the MIT License.
 */
package redis.clients.authentication.entraid;
```

File: entraid/src/main/java/redis/clients/authentication/entraid/EntraIDIdentityProvider.java [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+import com.microsoft.aad.msal4j.ManagedIdentityApplication;
+import com.microsoft.aad.msal4j.ManagedIdentityParameters;
+import redis.clients.authentication.core.IdentityProvider;
+import redis.clients.authentication.core.Token;
+
+public final class EntraIDIdentityProvider implements IdentityProvider {
+
+    private Supplier<IAuthenticationResult> resultSupplier;
+
+    public EntraIDIdentityProvider(ServicePrincipalInfo servicePrincipalInfo, Set<String> scopes,
```
Comment (tishun) [inline entraid/src/main/java/redis/clients/authentication/entraid/EntraIDIdentityProvider.java:23]: Its generally considered a bad idea to add logic in the constructor that could fail its creation.
An exception to this is classes such as URI where the contractor fails if the URI is malformed.

However here we don't validate the input parameters, but instead build up an app that connects remotely and there is a  potential that many of the steps could fail.

IMHO this needs to be either a separate step 
```java
provider.initialize();
```
or a lazy-loaded step to any of the public merhods (

```java
if(!initialized) { 
   initialize(); 
}
```

File: entraid/src/test/java/redis/clients/authentication/EndpointConfig.java [hunk 12]
```diff
@@ -0,0 +1 @@
+//
```
Comment (tishun) [inline entraid/src/test/java/redis/clients/authentication/EndpointConfig.java:1]: What do we need this file for?

**Review by tishun — APPROVED** [review]
Aaah a challenging review! 

Ali, thanks for all the work, this has been much harder than I personally anticipated.

I generally have few important comments, so none of this is blocking. The main things are:
- we are missing licensing statements, I've added them as suggestions, if you agree you can batch submit them
- we are missing a lot of documentation; I know we can add it later, but I always fear we will never come back to that; my personal minium is documentation for all public API, but in this case I would say please explain a few of the parts in the code that are a bit harder to understand (because of complexity)

File: core/src/main/java/redis/clients/authentication/core/TokenAuthConfig.java [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+        private int tokenRequestExecTimeoutInMs;
+        private int maxAttemptsToRetry;
+        private int delayInMsToRetry;
+
+        public T expirationRefreshRatio(float expirationRefreshRatio) {
+            this.expirationRefreshRatio = expirationRefreshRatio;
+            return (T) this;
+        }
+
+        public T lowerRefreshBoundMillis(int lowerRefreshBoundMillis) {
```
[the remark this answers is not in the document]
Comment (atakavci) [reply core/src/main/java/redis/clients/authentication/core/TokenAuthConfig.java:39]: not strongly against it. I think it is consistent that now we use ms everywhere when it comes to using/referring time. if we start playing with java types, better to use same approach everywhere like local dates etc. And i don't want that TBH. 
it always seems clear, easy to use and performant when we are not representing a human readable form of date/time/duration.

File: core/src/main/java/redis/clients/authentication/core/TokenManager.java [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+
+    public TokenManager(IdentityProvider identityProvider, TokenManagerConfig tokenManagerConfig) {
+        this.identityProvider = identityProvider;
+        this.tokenManagerConfig = tokenManagerConfig;
+    }
+
+    public void start(TokenListener listener, boolean blockForInitialToken) {
+
+        if (!started.compareAndSet(false, true)) {
+            throw new AuthXException("Token manager already started!");
```
[the remark this answers is not in the document]
Comment (atakavci) [reply core/src/main/java/redis/clients/authentication/core/TokenManager.java:39]: i ll think about this;
On one hand semantics of this method requires to be blocking if user prefers to. That makes multiple threads should be syncronised here. On the other hand the one who starts it knows the nature of its use and when to stop. Multiple start/stop means problem in that sense; when its exposed to a public API, who would stop should be the one who starts. Or isnt it ? 

Having said that, method signature might evolve to return if its an actual starter or not while letting multiple starts, may be..  with a safe initializer. let me work a bit on this.

File: core/src/main/java/redis/clients/authentication/core/TokenManager.java [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+        }
+        this.listener = listener;
+        ScheduledFuture<?> currentTask = scheduleNext(0);
+        scheduledTask = currentTask;
+        if (blockForInitialToken) {
+            try {
+                while (currentTask.get() == null) {
```
[the remark this answers is not in the document]
Comment (atakavci) [reply core/src/main/java/redis/clients/authentication/core/TokenManager.java:46]: thats my bad, i ll try to come up with comments as well as an easy to read coding espacially for actions running in parallel. There is a reasonable set of moving parts here.

File: core/src/main/java/redis/clients/authentication/core/TokenManager.java [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+            return (RuntimeException) unwrapped;
+        }
+        return new TokenRequestException(unwrapped, lastException);
+    }
+
+    public Token getCurrentToken() {
+        return currentToken;
+    }
+
+    public long calculateRenewalDelay(long expireDate, long issueDate) {
```
[the remark this answers is not in the document]
Comment (atakavci) [reply core/src/main/java/redis/clients/authentication/core/TokenManager.java:126]: right, i ll try to add clear explanations.

File: entraid/src/main/java/redis/clients/authentication/entraid/EntraIDIdentityProvider.java (hunk 11 above)
Comment (tishun) [reply entraid/src/main/java/redis/clients/authentication/entraid/EntraIDIdentityProvider.java:29]: Ali, https://stackoverflow.com/questions/6086334/is-it-good-practice-to-make-the-constructor-throw-an-exception this opinion may not be shared by the community, so please accept this only if you agree with me.

Comment (atakavci) [reply entraid/src/main/java/redis/clients/authentication/entraid/EntraIDIdentityProvider.java:29]: it s ok; i think what you mentioned is especially valid for static ctors,, i personally had some bad experience with static constructors in the past and i got goosebumps whenever i see them around. Though this is not the case, only down side of moving it to a later stage defers the initiation of identity provider context, which is ok i think.

File: core/src/main/java/redis/clients/authentication/core/AuthXException.java [hunk 17]
```diff
@@ -0,0 +1,12 @@
+package redis.clients.authentication.core;
+
+public class AuthXException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (atakavci) [reply core/src/main/java/redis/clients/authentication/core/AuthXException.java:9]: i ll add more documentation with another iteration

File: core/src/main/java/redis/clients/authentication/core/IdentityProvider.java [hunk 18]
```diff
@@ -0,0 +1,6 @@
+package redis.clients.authentication.core;
+
+public interface IdentityProvider {
+    
+    Token requestToken();
```
[the remark this answers is not in the document]
Comment (atakavci) [reply core/src/main/java/redis/clients/authentication/core/IdentityProvider.java:11]: i ll add more documentation with another iteration

File: core/src/main/java/redis/clients/authentication/core/IdentityProviderConfig.java [hunk 19]
```diff
@@ -0,0 +1,6 @@
+package redis.clients.authentication.core;
+
+public interface IdentityProviderConfig {
+
+    public IdentityProvider getProvider();
```
[the remark this answers is not in the document]
Comment (atakavci) [reply core/src/main/java/redis/clients/authentication/core/IdentityProviderConfig.java:11]: i ll add more documentation with another iteration

File: core/src/main/java/redis/clients/authentication/core/SimpleToken.java [hunk 20]
```diff
@@ -0,0 +1,57 @@
+package redis.clients.authentication.core;
+
+import java.util.Map;
+
+public class SimpleToken implements Token {
```
[the remark this answers is not in the document]
Comment (atakavci) [reply core/src/main/java/redis/clients/authentication/core/SimpleToken.java:11]: i ll add more documentation with another iteration

File: core/src/main/java/redis/clients/authentication/core/TokenManager.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (atakavci) [reply core/src/main/java/redis/clients/authentication/core/TokenManager.java:39]: will work on this before GA release

**Pull request merged**